cilium_eip_no_masquerade_agent/src/main.rs

```diff
@@ -98,36 +81,125 @@ impl k8s_controller::Context for Context {
-    pod_ip: Ipv4Addr,
-) -> Result<Vec<RuleMessage>, rtnetlink::Error> {
-    rules
-        .try_filter(|rule| {
-            future::ready(
-                rule.header.src_len == 32
-                    && rule.nlas.contains(&Nla::Source(pod_ip.octets().to_vec())),
-            )
-        })
-        .collect::<Vec<Result<RuleMessage, rtnetlink::Error>>>()
-        .await
-        .into_iter()
-        .collect()
+    async fn remove_finalizer(&self, pod: &Pod, pod_name: &str) {
+        // https://docs.rs/kube-runtime/latest/src/kube_runtime/finalizer.rs.html
+        let finalizer_index = pod
+            .finalizers()
+            .iter()
+            .enumerate()
+            .find(|(_, finalizer)| *finalizer == FINALIZER_NAME)
+            .map(|(i, _)| i);
+        if let Some(finalizer_index) = finalizer_index {
+            let pod_api: Api<Pod> =
+                Api::namespaced(self.kube_client.clone(), &pod.namespace().unwrap());
+            let finalizer_path = format!("/metadata/finalizers/{finalizer_index}");
+            pod_api
+                .patch::<Pod>(
+                    pod_name,
+                    &PatchParams::default(),
+                    &Patch::Json(json_patch::Patch(vec![
+                        // All finalizers run concurrently and we use an integer index
+                        // `Test` ensures that we fail instead of deleting someone else's finalizer
+                        PatchOperation::Test(TestOperation {
+                            path: finalizer_path.clone(),
+                            value: FINALIZER_NAME.into(),
+                        }),
+                        PatchOperation::Remove(RemoveOperation {
+                            path: finalizer_path,
+                        }),
+                    ])),
+                )
+                .await
+                .unwrap();
+        }
+    }
+
+    async fn wait_for_chain_to_exist(&self) -> Result<(), Box<dyn std::error::Error>> {
+        info!("Waiting for {CHAIN} chain to exist");
+        while !self.iptables.chain_exists(TABLE, CHAIN)? {
+            trace!("{CHAIN} does not exist, sleeping");
+            sleep(Duration::from_secs(1)).await;
+        }
+        Ok(())
+    }
+
+    async fn get_eni_indexes_for_existing_ip_rules(
+        &self,
+    ) -> Result<HashSet<u32>, rtnetlink::Error> {
+        let rules = self.ip_rule_handle.get(IpVersion::V4).execute();
+        rules
+            .try_filter(|rule| future::ready(rule.nlas.contains(&Nla::FwMask(0xf))))
+            .map_ok(|rule| {
+                rule.nlas
+                    .into_iter()
+                    .find_map(|nla| match nla {
+                        Nla::FwMark(i) => Some(i),
+                        _ => None,
+                    })
+                    .expect("There should always be an associated mark if there is a mask.")
+            })
+            .collect::<Vec<Result<u32, rtnetlink::Error>>>()
+            .await
+            .into_iter()
```

Comment:
don't think this extra iter/collect is needed, maybe just `.try_collect().await` and the return type is enough

Reply:
Ah, yeah, that's better. TIL about `try_collect`.